### ISSUE TITLE
[Bug]: Cannot delete from uncompressed chunk if previous chunks are still compressed

### ISSUE THREAD

**Issue by ScottLangridge** [timescale/timescaledb#4987]
### What type of bug is this?

Unexpected error

### What subsystems and features are affected?

Compression

### What happened?

If you try to delete from a partially compressed hypertable (where you have already decompressed the chunks you need to delete from in advance) you get an error: `cannot update/delete rows from chunk "FIRST COMRPESSED CHUNK IN THE TABLE" as it is compressed`.

This appears to be because Timescale attempts to scan each chunk as part of the delete _before_ checking whether they are compressed, causing an error on the first uncompressed chunk it hits.

For anyone else experiencing this, this can be worked around by deleting directly from the chunk itself. Eg: `DELETE FROM _timescaledb_internal._hyper_13_40_chunk WHERE x=y;`

### TimescaleDB version affected

2.7.0

### PostgreSQL version used

14.4

### What operating system did you use?

Debian 11 x86_64

### What installation method did you use?

Docker

### What platform did you run on?

On prem/Self-hosted

### Relevant log output and stack trace

```bash
project_beta_development=# CREATE TABLE test (
  time TIMESTAMP,
  value INTEGER
);
CREATE TABLE

project_beta_development=# INSERT INTO test
VALUES
  ('2022-01-01 00:00',0),
  ('2022-01-01 00:01',1),
  ('2022-01-01 00:02',2),
  ('2022-01-01 00:03',3),
  ('2022-01-01 00:04',4),
  ('2022-01-01 00:05',5),
  ('2022-01-01 00:06',6),
  ('2022-01-01 00:07',7),
  ('2022-01-01 00:08',8);
INSERT 0 9

project_beta_development=# SELECT create_hypertable('test', 'time', chunk_time_interval => INTERVAL '3 mins', migrate_data => TRUE);
NOTICE:  adding not-null constraint to column "time"
DETAIL:  Time dimensions cannot have NULL values.
NOTICE:  migrating data to chunks
DETAIL:  Migration might take a while depending on the amount of data.
 create_hypertable
-------------------
 (1,public,test,t)
(1 row)

project_beta_development=# ALTER TABLE test SET (timescaledb.compress);
ALTER TABLE

project_beta_development=# SELECT compress_chunk(i) FROM show_chunks('test') i;
             compress_chunk
----------------------------------------
 _timescaledb_internal._hyper_1_1_chunk
 _timescaledb_internal._hyper_1_2_chunk
 _timescaledb_internal._hyper_1_3_chunk
(3 rows)

project_beta_development=# SELECT tableoid::regclass AS name
FROM test
WHERE value = 6
GROUP BY tableoid;
                  name
----------------------------------------
 _timescaledb_internal._hyper_1_3_chunk
(1 row)

project_beta_development=# CREATE TABLE chunks_to_decompress AS (
  SELECT tableoid::regclass AS name
  FROM test
  WHERE value = 6
  GROUP BY tableoid
);
SELECT 1

project_beta_development=# SELECT decompress_chunk(name) FROM chunks_to_decompress;
            decompress_chunk
----------------------------------------
 _timescaledb_internal._hyper_1_3_chunk
(1 row)

project_beta_development=# DELETE FROM test
WHERE value = 6;
ERROR:  cannot update/delete rows from chunk "_hyper_1_1_chunk" as it is compressed
```

### How can we reproduce the bug?

```bash
-- Set up hypertable
CREATE TABLE test (
  time TIMESTAMP,
  value INTEGER
);

INSERT INTO test
VALUES
  ('2022-01-01 00:00',0),
  ('2022-01-01 00:01',1),
  ('2022-01-01 00:02',2),
  ('2022-01-01 00:03',3),
  ('2022-01-01 00:04',4),
  ('2022-01-01 00:05',5),
  ('2022-01-01 00:06',6),
  ('2022-01-01 00:07',7),
  ('2022-01-01 00:08',8);

SELECT create_hypertable('test', 'time', chunk_time_interval => INTERVAL '3 mins', migrate_data => TRUE);

ALTER TABLE test SET (timescaledb.compress);

SELECT compress_chunk(i) FROM show_chunks('test') i;

-- Find chunk with data to delete
SELECT tableoid::regclass AS name
FROM test
WHERE value = 6
GROUP BY tableoid;

-- Decompress the chunk
CREATE TABLE chunks_to_decompress AS (
  SELECT tableoid::regclass AS name
  FROM test
  WHERE value = 6
  GROUP BY tableoid
);

SELECT decompress_chunk(name) FROM chunks_to_decompress;

-- Attempt to delete the data
DELETE FROM test
WHERE value = 6;
```
```

**Comment by konskov** [comment]
hi @ScottLangridge thank you for reporting the issue with reproduction steps, and glad to know you have found a workaround!
We are currently working on supporting DELETE/UPDATE on compressed chunk, so that work will fix this issue too.

**Comment by ScottLangridge** [comment]
hi @konskov, that's great to hear! Thank you. 

I know you can never be 100% sure with these things, but is there a rough estimate for when that is likely to be rolled out?

**Comment by konskov** [comment]
@ScottLangridge the rough estimate is sometime in the beginning of next year.

**Comment by mkindahl** [comment]
@ScottLangridge Looking at the plan for the delete query, I get this (after decompressing the last chunk):

```
mats=# explain verbose DELETE FROM test WHERE value = 6;
                                                              QUERY PLAN                                                              
--------------------------------------------------------------------------------------------------------------------------------------
 Custom Scan (HypertableModify)  (cost=0.00..3.13 rows=3 width=10)
   ->  Delete on public.test  (cost=0.00..3.13 rows=3 width=10)
         Delete on _timescaledb_internal._hyper_1_1_chunk test_1
         Delete on _timescaledb_internal._hyper_1_2_chunk test_2
         Delete on _timescaledb_internal._hyper_1_3_chunk test_3
         ->  Append  (cost=0.00..3.13 rows=3 width=10)
               ->  Custom Scan (CompressChunkDml) on _timescaledb_internal._hyper_1_1_chunk test_1  (cost=0.00..1.04 rows=1 width=10)
                     Output: test_1.tableoid, test_1.ctid
                     ->  Seq Scan on _timescaledb_internal._hyper_1_1_chunk test_1  (cost=0.00..1.04 rows=1 width=10)
                           Output: test_1.tableoid, test_1.ctid
                           Filter: (test_1.value = 6)
               ->  Custom Scan (CompressChunkDml) on _timescaledb_internal._hyper_1_2_chunk test_2  (cost=0.00..1.04 rows=1 width=10)
                     Output: test_2.tableoid, test_2.ctid
                     ->  Seq Scan on _timescaledb_internal._hyper_1_2_chunk test_2  (cost=0.00..1.04 rows=1 width=10)
                           Output: test_2.tableoid, test_2.ctid
                           Filter: (test_2.value = 6)
               ->  Seq Scan on _timescaledb_internal._hyper_1_3_chunk test_3  (cost=0.00..1.04 rows=1 width=10)
                     Output: test_3.tableoid, test_3.ctid
                     Filter: (test_3.value = 6)
(19 rows)
```

In other words, it tries to scan all chunks of the table and will therefore hit the compressed chunks as well.

**Comment by ScottLangridge** [comment]
Thank you for confirming that @mkindahl, I had suspected that something like that would be the case. 

The workaround that I mentioned in the original report does the job to avoid it, so I'll leave the code as it is until the new functionality that @konskov mentioned is rolled out, then tidy up.

**Comment by Kraust** [comment]
The above suggestion also works with Update as well. I've been able to implement update on a partially compressed hypertable by selecting the chunks that have the data I want to update and performing the update operations on the chunks instead of the table.

**Comment by invernizzie** [comment]
@fabriziomello I see you closed this on Jan 3. Previously, @konskov mentioned it was related to the ability to DELETE from compressed chunks. Has that shipped as well? Is there an MR you can reference? Thank you.

**Comment by mkindahl** [comment]
Seems to have been closed by accident, we have not yet fixed chunk pruning for update and delete. Reopening.

**Comment by sb230132** [comment]
Hi @ScottLangridge 
Chunk pruning cannot be done, because we dont know if this condition value = 6 holds true or false unless we decompress the chunks. This we scan all chunks.

However the DELETE issue is fixed In version 2.11.0.

```
test=*# explain SELECT * FROM test WHERE value = 6;
                                            QUERY PLAN
--------------------------------------------------------------------------------------------------
 Append  (cost=0.02..7069.51 rows=1400001 width=12)
   ->  Custom Scan (DecompressChunk) on _hyper_1_1_chunk  (cost=0.02..17.00 rows=700000 width=12)
         Filter: (value = 6)
         ->  Seq Scan on compress_hyper_2_4_chunk  (cost=0.00..17.00 rows=700 width=84)
   ->  Custom Scan (DecompressChunk) on _hyper_1_2_chunk  (cost=0.02..17.00 rows=700000 width=12)
         Filter: (value = 6)
         ->  Seq Scan on compress_hyper_2_5_chunk  (cost=0.00..17.00 rows=700 width=84)
   ->  Seq Scan on _hyper_1_3_chunk  (cost=0.00..35.50 rows=1 width=12)
         Filter: (value = 6)
(9 rows)

test=*# SELECT * FROM test WHERE value = 6;
        time         | value
---------------------+-------
 2022-01-01 00:06:00 |     6
(1 row)

test=*# DELETE FROM test WHERE value = 6;
DELETE 1
```
